src/goto-programs/remove_function_pointers.cpp

```diff
@@ -359,6 +373,117 @@ void remove_function_pointerst::remove_function_pointer(
   remove_function_pointer(goto_program, function_id, target, functions);
 }
 
+
+/// For a call site, return a list of potential targets.
+/// param goto_program: the goto_program containing the call site
+/// param target: the actual callsite
+/// return: A list of potential function identifiers that can be
+///         the targets of this callsite.
+remove_function_pointerst::possible_fp_targetst
+remove_function_pointerst::list_potential_targets(
+  goto_programt &goto_program,
+  goto_programt::targett target)
+{
+  remove_function_pointerst::possible_fp_targetst targets;
+  bool found_functions;
+
+  const code_function_callt &code =
+    to_code_function_call(target->code);
+
+  const exprt &function = code.function();
+
+  // this better have the right type
+  code_typet call_type = to_code_type(function.type());
+
+  // refine the type in case the forward declaration was incomplete
+  if(call_type.has_ellipsis() &&
+     call_type.parameters().empty())
+  {
+    call_type.remove_ellipsis();
+    forall_expr(it, code.arguments())
+      call_type.parameters().push_back(
+        code_typet::parametert(it->type()));
+  }
+
+  const exprt &pointer = function.op0();
+  functionst functions;
+  does_remove_constt const_removal_check(goto_program, ns);
+  const auto does_remove_const = const_removal_check();
+  if(does_remove_const.first)
+  {
+    log.warning().source_location = does_remove_const.second;
+    log.warning() << "Cast from const to non-const pointer found, only worst case"
+                  << " function pointer removal will be done." << messaget::eom;
+    found_functions = false;
+  }
+  else
+  {
+    remove_const_function_pointerst fpr(
+      log.get_message_handler(), ns, symbol_table);
+
+    found_functions = fpr(pointer, functions);
+
+    // if found_functions is false, functions should be empty
+    // however, it is possible for found_functions to be true and functions
+    // to be empty (this happens if the pointer can only resolve to the null
```

Comment:
I think this interface needs to change. IMHO we should make `operator()` return an `optionalt<functionst>` to avoid this mess of having to check the combined value of a return value an a modified argument.